### PR TITLE
Injection maps handle Supports

### PR DESCRIPTION
**Pull request by mikeheaton-stripe**
This PR makes the `Injection` transforms take care of the `Support` of distributions.

### CODE REVIEW

**Comment by avi-stripe** [comment]
:+1: to what's here so far

File: rainier-core/src/main/scala/com/stripe/rainier/core/Injection.scala [hunk 1]
```diff
@@ -6,19 +6,21 @@ import com.stripe.rainier.unused
...
   * Injective transformations
   */
-trait Injection { self =>
+private[rainier] trait Injection { self =>
   def forwards(x: Real): Real
   def backwards(y: Real): Real
   def isDefinedAt(@unused y: Real): Real = Real.one
   def requirements: Set[Real]
-
+  def supportTransform(supp: Support): Support
```
Comment (avi-stripe) [inline rainier-core/src/main/scala/com/stripe/rainier/core/Injection.scala]: can we call this method `transformSupport` instead?

File: rainier-core/src/main/scala/com/stripe/rainier/core/Injection.scala [hunk 2]
```diff
@@ -35,37 +37,56 @@ trait Injection { self =>
...
-  * Class to scale a distribution under multiplication by a scale factor
+  * Class to scale a distribution under multiplication by a positive scale factor.
+  * We assume that (a > 0).
   */
 final case class Scale(a: Real) extends Injection {
-  private val lj = a.log * -1
   def forwards(x: Real): Real = x * a
   def backwards(y: Real): Real = y / a
-  def logJacobian(y: Real): Real = lj
+  def logJacobian(y: Real): Real = a.log * -1
```
Comment (avi-stripe) [inline rainier-core/src/main/scala/com/stripe/rainier/core/Injection.scala]: Is there a reason you got rid of the shared object for this? I try to be careful about not over-allocating where possible.

File: rainier-core/src/main/scala/com/stripe/rainier/core/Injection.scala [hunk 3]
```diff
@@ -35,37 +37,56 @@ trait Injection { self =>
...
-  //this is rarely important because it depends solely on y, which is usually data and not parameters
   def logJacobian(y: Real): Real = y.log * -1
 
-  override def isDefinedAt(y: Real): Real =
-    y > 0
+  override def isDefinedAt(y: Real): Real = y > 0
   val requirements: Set[Real] = Set.empty
+
+  def supportTransform(supp: Support): Support = supp match {
+    case UnboundedSupport         => UnboundedSupport
```
Comment (avi-stripe) [inline rainier-core/src/main/scala/com/stripe/rainier/core/Injection.scala]: shouldn't this end up as `BoundedBelowSupport(Real.zero)`?

File: rainier-core/src/main/scala/com/stripe/rainier/core/Continuous.scala [hunk 4]
```diff
@@ -152,7 +154,7 @@ object Exponential {
   * A Beta distribution with expectation `a/(a + b)` and variance `ab/((a + b)^2 (1 + a + b))`.
   */
 final case class Beta(a: Real, b: Real) extends StandardContinuous {
-  val support = OpenUnitSupport
+  val support = new BoundedSupport(Real.zero, Real.one)
```
Comment (avi-stripe) [inline rainier-core/src/main/scala/com/stripe/rainier/core/Continuous.scala]: don't need `new` for a case class

Comment (mikeheaton-stripe) [reply rainier-core/src/main/scala/com/stripe/rainier/core/Continuous.scala]: Done

File: rainier-core/src/main/scala/com/stripe/rainier/core/Injection.scala (hunk 1 above)
Comment (mikeheaton-stripe) [reply rainier-core/src/main/scala/com/stripe/rainier/core/Injection.scala]: Changed

File: rainier-core/src/main/scala/com/stripe/rainier/core/Injection.scala (hunk 2 above)
Comment (mikeheaton-stripe) [reply rainier-core/src/main/scala/com/stripe/rainier/core/Injection.scala]: `lj` wasn't being used anywhere else so I thought it should be inlined... what do you mean by over-allocating here?

Comment (mikeheaton-stripe) [reply rainier-core/src/main/scala/com/stripe/rainier/core/Injection.scala]: Ah, putting it in a `val` will make it calculate only once, right?

Comment (avi-stripe) [reply rainier-core/src/main/scala/com/stripe/rainier/core/Injection.scala]: exactly

**Pull request merged**